### DISCUSSION TITLE
Measuring the water impact of software

### DISCUSSION THREAD

**Discussion by jawache** [Green-Software-Foundation/hack#82] [Develop an Idea]
This is an idea for measuring the water impact of software, and would fit into the beyond carbon prize category of the hackathon. If you would like to propose this or some version of this as a hackathon project you can read up how to here: https://github.com/Green-Software-Foundation/hack/wiki/Participant-Guide

### Calculating how much water your software consumes

The primary use of water in data centers for software is in cooling and as such it's proportional to the energy usage of yoru application.

It should therefore be possible to **induce** and observation of energy into an impact of water.

### WUE

"Water Usage Effectiveness (WUE) is a critical metric for assessing the water efficiency of data centers. It specifically measures the amount of water used to cool the facility’s equipment.

WUE is calculated as the ratio of the data center’s annual water use (in liters) to the energy consumed by its IT computing equipment, measured in kilowatt hours (kWh). Thus, WUE is expressed in liters per kilowatt-hour (L/kWh). **According to Meta Platforms (formerly Facebook), the industry’s average WUE was most recently 1.80 L/kWh**, indicating that 1.80 liters of water were used for every kilowatt-hour of electricity consumed."
 https://dgtlinfra.com/data-center-water-usage/#:~:text=Water%20Usage%20Effectiveness%20(WUE)%20is,kilowatt%2Dhour%20of%20electricity%20consumed.

Using the above WUE we can therefore estimate the amount of water your application consumes by multiplying the energy consumed by 1.8.

That gives you water usage, it's a very rough estimate since WUE would be different for each cloud providers, each data center and each region. Some cloud providers provide this information at varying levels of granularity.

**Proposal: WUE plugin**
- Maintains a data set of WUE for each Cloud, DC.
- When there is an input of **cloud/vendor** and **cloud/region** it outputs cloud/region-wue which is the WUE for that region.
- If there is no data for that region it can output a default such as 1.8 above or some other measure.
- This could be a separate standalone plugin or an enhancement to https://github.com/Green-Software-Foundation/if-plugins/blob/main/src/lib/cloud-instance-metadata/README.md.

### Applying it to calculate a Software Water Intensity

That doesn't give you much to go on, it's just a total and as a metric for software developer we can explore something like the [SCI](https://sci.greensoftware.foundation/). The SCI is a rate of carbon, **carbon per X**, SWI could be a water (in L) per X. 

There is an existing SCI plugin which you can model this on, OR even just hack a solution by treating water as carbon and using the SCI plugin to compute the Water per X using it's existing algorithms.

**Proposal: SWI plugin**
- A plugin mirroring the logic of the current SCI plugin: https://github.com/Green-Software-Foundation/if-plugins/blob/main/src/lib/sci/README.md but which returns a software water intensity metric rather than a carbon metric.
- inputs: water (in L) 
- config: functional-unit (what to divide water by, can be time or another field in the inputs like number-of-requests, same as existing SCI plugin)
- outputs: swi (Water per Functional Unit)

### Another approach to surface real-world human implications.

There is a data set which gives an amount of water available per capita per country https://ourworldindata.org/grapher/per-capita-renewable-freshwater-resources

This is in m3 per person, it should then be possible using this data to calculate some form of human impact and how it differs region by region.

E,.g In South Asia the total total amount of yearly renewable water available per person is 1,131,000 Liters. So If you're all used 100 L per say a 1hr duration. Then over the year it would consume 867,000 L of water. So it would consume 0.76 of a humans available worth of water in that region. If you ran the same software in Easter Europe it would only consume 0.04 worth of a humans available water.

Whether we use that approach or another, some way of linking software water usage to the impact in a particular region would be an ideal endpoint.

If there are data sources which contain more granular data that might be much more useful since we can then both be more accurate and also provide more interesting solutions, moving to a DC a little further away might change things in very interesting ways.

If you can't move compute this kind of metric would provide increased incentivisation to make software more energy efficient, if you can't move the compute, make the compute use less energy so it consumes less water and has less human impact.

**Proposal: Water Displacement Plugin**
- A plugin which given some water consumption use for an application (Water in L) computes for that region how many humans worth of water will your application consume.
- inputs: Water in L
- data source: Perhaps this or something more granular https://ourworldindata.org/grapher/per-capita-renewable-freshwater-resources
outputs: humans/water (or some better term)

### Comparing the carbon impact to the water impact

Bonus points for comparing carbon impact and water impact. A single pipeline of plugins could compute both, one set outputing carbon and sci scores and other plugin(s) outputting water and SWI scores. It's been proposed that perhaps making solutions. more water efficient might forice choices to make them less carbon efficient. Surfacing this with a common set of plugins would push this conversation forward in useful ways.

**Comment by maddendeavor** [comment]
@jawache Hi!  A colleague and I were discussing this idea of the water impact plugin and wanted to try to run with it for the hackathon.  I am not quite sure if this was an idea for participants to work on, or more of a backlog item for the GSF team?  If available, we will submit a project based on at least part of this.  Thanks for all the detail and the great idea!

**Reply by jawache** [reply]
That's wonderful @maddendeavor, it's totally something I threw out there and hoped someone would pick up! Please submit it as a hackathon project 😁

**Comment by ysmin-g** [comment]
@jawache Hi there! our group is also interested in studying parts of this idea for the competition.

**Reply by jawache** [reply]
That's wonderful! There are lots here to explore, water is fast becoming the next most important environmental impact, we need people like yourself thinking about it!

**Comment by derekharmon** [comment]
I find this proposed model for measuring software's water impact has 2 significant flaws:

### 1. Water Impact of Electricity Generation

Focusing solely on the water consumption necessary to cool a data center excludes the water consumption necessary for electricity generation used by the data center. 

- Several carbon-free energy sources such as hydropower and nuclear consume **substantial** quantities of water.
- Models of their water use can **vary widely** depending on very site-specific configurations (single- vs. multi-reservoir, river basin spatial allocations among multiple hydroelectric plants, temperature, humidity, elevation).  Net water use per MWh can range from as little as 0.2 to as much as 140 cubic meters ([Bakken, et al. (2017)](https://doi.org/10.1002/gch2.201600018)).
- Information about the data center's energy supply mix is necessary to apportion this water impact.

To the point of your suggestion, "It's been proposed that perhaps making solutions more water efficient might force choices to make them less carbon efficient."  Certainly, making them more carbon-efficient can make them less water-efficient when the electricity necessary to run the data center comes from a mix of nuclear, hydro and geothermal power, but measuring only the impact of data center water-cooling overlooks this factor.

#### Google Cloud Platform Low-Carbon Data Centers

Hydroelectric power generation is plentiful in The Dalles, Oregon (see [BPA Hydro History (2023)](https://www.wascoelectric.com/hydro-history/)) where the Google Cloud Platform `us-west-1` data center is located, and is largely responsible for that region's 89% Carbon-Free Energy ([67 gCO2eq/kWh](https://cloud.google.com/sustainability/region-carbon)) measure.  Other GCP data centers such as in the `northamerica-northeast1` region [outside of Montreal](https://montrealgazette.com/business/local-business/google-to-set-up-735-million-data-centre-in-beauharnois) achieves its 100% Carbon-Free Energy with an energy mix from supplier [Hydro-Québec](https://montrealgazette.com/tag/hydro-quebec/) which includes hydroelectric, nuclear and wind farms.

### 2. Water Impact of Thermal Power Management

Focusing solely on the water consumption necessary to cool a data center excludes the impact of TPM at the software and operating system level, among other design choices which can make trade-offs that reduce operating temperature.

The active compute components that primarily require this cooling are:

- CPUs (GPUs. TPUs, compute accelerators)
- DRAM
- SSDs (storage)
- network interfaces

Water cooling is necessary to keep these components within their operating temperature ranges. The heat produced is proportional to the activity by which the software exercises these components, e.g., the following software design choices should reduce the power consumption -> heat generation -> water cooling needs of software running on a given rack-mounted unit.

- The greater the data locality possible (registers >> L1 cache SRAM >> Ln cache SRAM >> DRAM >> SSD >> networking) then the lower the power consumption due to data movement, although this typically implies CPU (compute) power consumption is high and the compute is where the greatest heat is generated.
- Processing In-/Near-Memory (PIM) architectures reduce data movement and consequently, the power consumption and heat generation due to data movement.
- Cores within a CPU package can be put to varying levels of sleep (Core and Package C-states) and thermal power management modes (see Intel cTDP and LPM), powering down physically-adjacent cores and/or levels of internal cache based on what the software requires. These measures can reduce power consumption below the CPU's stated total power rating (Intel TDP, AMD ACP; these manufacturers' power ratings appear outwardly similar but are not calculated equivalently apples-to-apples) leading to lower temperatures and lower water demand for cooling.

#### The Trade-Offs of Monolithic vs. Distributed Software Systems 

Is the water impact of 2 `n2-standard` VMs where one VM does processing and the second VM hosts a Postgres 16 database server (with its intrinsic data movement and remoteness of storage) compared to 1 larger `n4-standard` VM performing processing with a local SqlLite 3 database better or worse?   I don't see how a data center average WUE captures the effect on a particular software solution's water impact of these design choices.

### Conclusion

Baaken, et al. (2017) describes several research challenges that still confound our understanding and modeling of the water impact for hydropower generation which is a necessary input here (as well as other energy sources renewable and otherwise) that can't be ignored in modeling software's water impact.  Substantially better integrated information (it may exist, but in disparate, private, siloed sources) on energy supplier mix, weather, and GIS is necessary to make an accurate model of software water impact.  Analyses based on measures at the data center level are too coarse-grained to make informed software design decisions that reduce water impact.  This proposed model may produce an estimated water impact, but I think its' margin-of-error is such as to be practically useless.

**Reply by jawache** [reply]
Thanks Derek. Do you have any data sources that people could use to create a model of this? The only reason we can take a carbon grid mix into consideration for carbon measurement is because there's now data providers or freely available data which gives you carbon per kilowatt hour at different temporal granularities for every grid around the world.

What I think you're proposing is to create a similar data provider source which gives you litres of water per kilowatt hour, per grid, per some temporal granularity. Which I totally agree if it existed, could really help in creating a more accurate value.

If there was even a data source where the yearly averages per grid of litres of water per kilowatt hour, that actually might be quite easy to initially create a plug-in for.
